lib/Sema/CSSolver.cpp

```diff
@@ -1358,6 +1358,221 @@ static Constraint *selectBestBindingDisjunction(
+              cs.getRepresentative(secondTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep)) {
+            // See if we can actually determine what the underlying
+            // type is.
+            Type fixedTy;
+            auto firstTy = constraint->getFirstType();
+            if (!firstTy->is<TypeVariableType>()) {
+              fixedTy = firstTy;
+            } else {
+              fixedTy =
+                cs.getFixedType(firstTy->castTo<TypeVariableType>());
+            }
+            if (fixedTy && fixedTy->getOptionalObjectType())
+              worklist.push_back(fixedTy->getOptionalObjectType());
+          }
+        }
+        break;
+      }
+
+      case ConstraintKind::KeyPathApplication:
+      case ConstraintKind::KeyPath: {
+        auto firstTy = constraint->getFirstType();
+        if (firstTy->is<TypeVariableType>()) {
+          auto otherRep =
+              cs.getRepresentative(firstTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep))
+            worklist.push_back(constraint->getThirdType());
+        }
+        break;
+      }
+
+      case ConstraintKind::BindToPointerType:
+      case ConstraintKind::ValueMember:
+      case ConstraintKind::UnresolvedValueMember:
+      case ConstraintKind::Disjunction:
+      case ConstraintKind::CheckedCast:
+      case ConstraintKind::OpenedExistentialOf:
+      case ConstraintKind::ApplicableFunction:
+      case ConstraintKind::BindOverload:
+      case ConstraintKind::FunctionInput:
+      case ConstraintKind::FunctionResult:
+      case ConstraintKind::SelfObjectOfProtocol:
+      case ConstraintKind::ConformsTo:
+        break;
+      }
+    }
+  }
+
+  return false;
+}
+
+// Check to see if we know something about the types of all arguments
+// in the given function type.
+static bool haveTypeInformationForAllArguments(AnyFunctionType *fnType,
+                                               ConstraintSystem &cs) {
+  for (auto param : fnType->getParams())
+    if (!havePotentialTypesOrLiteralConformances(param.getType(), cs))
+      return false;
+
+  return true;
+}
+
+// Given a type variable representing the RHS of an ApplicableFunction
+// constraint, attempt to find the disjunction of bind overloads
+// associated with it. This may return null in cases where have not
+// yet created a disjunction because we need to resolve a base type,
+// e.g.: [1].map{ ... } does not have a disjunction until we decide on
+// a type for [1].
+static Constraint *getUnboundBindOverloadDisjunction(TypeVariableType *tyvar,
+                                                     ConstraintSystem &cs) {
+  auto *rep = cs.getRepresentative(tyvar);
+  assert(!cs.getFixedType(rep));
+
+  llvm::SetVector<Constraint *> disjunctions;
+  cs.getConstraintGraph().gatherConstraints(
+      rep, disjunctions, ConstraintGraph::GatheringKind::EquivalenceClass,
+      [](Constraint *match) {
+        return match->getKind() == ConstraintKind::Disjunction;
```

Comment:
It looks like logic from the `for` loop at the end of this method could be moved into filtering callback, because that test doesn't really require any additional information besides what disjunction has...

Reply:
Yeah this normally returns zero or one matches, but you're right I can just fold the logic in. I'll push a change for that as well.